include/nbl/video/SPhysicalDeviceLimits.h

```diff
@@ -0,0 +1,497 @@
+#ifndef _NBL_VIDEO_S_PHYSICAL_DEVICE_LIMITS_H_INCLUDED_
+#define _NBL_VIDEO_S_PHYSICAL_DEVICE_LIMITS_H_INCLUDED_
+
+#include <type_traits>
+#include "nbl/asset/utils/IGLSLCompiler.h" // asset::IGLSLCompiler::E_SPIRV_VERSION
+
+
+namespace nbl::video
+{
+
+struct SPhysicalDeviceLimits
+{
+    /* Vulkan 1.0 Core  */
+    uint32_t maxImageDimension1D;
+    uint32_t maxImageDimension2D;
+    uint32_t maxImageDimension3D;
+    uint32_t maxImageDimensionCube;
+    uint32_t maxImageArrayLayers;
+    uint32_t maxBufferViewSizeTexels;
+    uint32_t maxUBOSize;
+    uint32_t maxSSBOSize;
+    //uint32_t              maxPushConstantsSize;
+    //uint32_t              maxMemoryAllocationCount;
+    //uint32_t              maxSamplerAllocationCount;
+    //VkDeviceSize          bufferImageGranularity;
+    //VkDeviceSize          sparseAddressSpaceSize;
+    //uint32_t              maxBoundDescriptorSets;
+    //uint32_t              maxPerStageDescriptorSamplers;
+    //uint32_t              maxPerStageDescriptorUniformBuffers;
+    uint32_t maxPerStageDescriptorSSBOs;
+    //uint32_t              maxPerStageDescriptorSampledImages;
+    //uint32_t              maxPerStageDescriptorStorageImages;
+    //uint32_t              maxPerStageDescriptorInputAttachments;
+    //uint32_t              maxPerStageResources;
+    //uint32_t              maxDescriptorSetSamplers;
+    uint32_t maxDescriptorSetUBOs;
+    uint32_t maxDescriptorSetDynamicOffsetUBOs;
+    uint32_t maxDescriptorSetSSBOs;
+    uint32_t maxDescriptorSetDynamicOffsetSSBOs;
+    uint32_t maxDescriptorSetImages;
+    uint32_t maxDescriptorSetStorageImages;
+    //uint32_t              maxDescriptorSetInputAttachments;
```

Comment:
ok so when you have 

### maxPerStageDescriptor

```
```

### maxDescriptorSet

```
Samplers = GL_MAX_TEXTURE_IMAGE_UNITS
UBOs = 
DynamicOffsetUBOs = 
SSBOs = GL_MAX_COMBINED_SHADER_STORAGE_BLOCKS
DynamicOffsetSSBOs = 
Images = 0 // because OpenGL doesn't support separable sampler and image
```

Reply:
Can you describe the exact logic of `maxPerStageResources` I didn't get why it looks like this. then I can also put a comment in code

Reply:
> maxPerStageResources is the maximum number of resources that can be accessible to a single shader stage in a pipeline layout. Descriptors with a type of VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER, VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC, or VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT count against this limit. Only descriptors in descriptor set layouts created without the VK_DESCRIPTOR_SET_LAYOUT_CREATE_UPDATE_AFTER_BIND_POOL_BIT bit set count against this limit. For the fragment shader stage the framebuffer color attachments also count against this limit.

```cpp
const uint32_t maxFragmentStageResources = GL_MAX_COMBINED_SHADER_OUTPUT_RESOURCES+ // VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC, For the fragment shader stage the framebuffer color attachments also count against this limit.
    GL_MAX_FRAGMENT_SHADER_UNIFORM_BLOCKS+ // VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC
    GL_MAX_TEXTURE_IMAGE_UNITS; // VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER
m_limits.maxPerStageResources = maxFragmentStageResources;

constexpr for (auto stage : {COMPUTE,VERTEX,TESS_CONTROL,TESS_EVAL,GEOMETRY})
{
    const uint32_t maxStageResources = GL_MAX_##stage##_TEXTURE_IMAGE_UNITS+ // VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER
        GL_MAX_##stage##_UNIFORM_BLOCKS+ // VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC
        GL_MAX_##stage##_SHADER_STORAGE_BLOCKS+ // VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC
        GL_MAX_##stage##_IMAGE_UNIFORMS; // VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER
}
```

Reply:
I think `max(maxStageResources ,m_limits.maxPerStageResources);` should be `min` ?! because it means the max resources we can create per stage so It's kinda a min of maxes